src/annotator/anchoring/range-util.js

```diff
@@ -0,0 +1,227 @@
+import $ from 'jquery';
+
+/**
+ * A simple XPath evaluator using jQuery which can evaluate queries of
+ */
+function simpleXPathJQuery(nodes, relativeRoot) {
+  const paths = nodes.map((index, node) => {
+    let path = '';
+    let elem = node;
+    while (elem.nodeType === Node.ELEMENT_NODE && elem !== relativeRoot) {
+      let tagName = elem.tagName.replace(':', '\\:');
+      let idx = $(elem.parentNode).children(tagName).index(elem) + 1;
+      path = '/' + elem.tagName.toLowerCase() + `[${idx}]` + path;
+      elem = elem.parentNode;
+    }
+    return path;
+  });
+  return paths.get();
+}
+
+function findChild(node, type, index) {
+  if (node.hasChildNodes()) {
+    throw new Error('XPath error: node has no children!');
+  }
+  const children = node.childNodes;
+  let found = 0;
+  for (let i = 0; i < children.length; i++) {
+    const child = children[i];
+    let name = getNodeName(child);
+    if (name === type) {
+      found += 1;
+      if (found === index) {
+        return child;
+      }
+    }
+  }
+  throw new Error('XPath error: wanted child not found.');
+}
+
+/**
+ * Get the node name for use in generating an xpath expression.
+ */
+function getNodeName(node) {
+  const nodeName = node.nodeName.toLowerCase();
+  let result = nodeName;
+  if (nodeName === '#text') {
+    result = 'text()';
+  }
+  if (nodeName === '#comment') {
+    result = 'comment()';
+  }
+  if (nodeName === '#cdata-section') {
+    result = 'cdata-section()';
+  }
+  return result;
+}
+
+/**
+ * Get the index of the node as it appears in its parent's child list
+ */
+function getNodePosition(node) {
+  let pos = 0;
+  let tmp = node;
+  while (tmp) {
+    if (tmp.nodeName === node.nodeName) {
+      pos += 1;
+    }
+    tmp = tmp.previousSibling;
+  }
+  return pos;
+}
+
+/**
+ * A simple XPath evaluator using only standard DOM methods which can
+ * evaluate queries of the form /tag[index]/tag[index].
+ */
+function simpleXPathPure(nodes, relativeRoot) {
+  function getPathSegment(node) {
+    const name = getNodeName(node);
+    const pos = getNodePosition(node);
+    return `${name}[${pos}]`;
+  }
+
+  let rootNode = relativeRoot;
+
+  function getPathTo(node) {
+    let xpath = '';
+    let elem = node;
+    while (elem !== rootNode) {
+      if (!elem) {
+        throw new Error(
+          'Called getPathTo on a node which was not a descendant of @rootNode. ' +
+            rootNode
+        );
+      }
+      xpath = getPathSegment(elem) + '/' + xpath;
+      elem = elem.parentNode;
+    }
+    xpath = '/' + xpath;
+    xpath = xpath.replace(/\/$/, '');
+    return xpath;
+  }
+  const paths = nodes.map((index, node) => {
+    return getPathTo(node);
+  });
+  return paths.get();
+}
+
+/**
+ * Public: Flatten a nested array structure
+ */
+function flatten(array) {
```

Comment:
We could eventually replace this with `Array.prototype.flat` from ES2019, we'd need to set up a polyfill for older browsers for that though. In the interim you could replace `$.isArray` with `Array.isArray` below.